packages/central-server/src/apiV2/surveyResponses/saveResponsesToDatabase.js

```diff
@@ -0,0 +1,139 @@
+import { getTimezoneNameFromTimestamp, stripTimezoneFromDate, yup } from '@tupaia/utils';
+import keyBy from 'lodash.keyby';
+import { upsertAnswers } from '../../dataAccessors';
+
+async function getRecordsByCode(model, codes) {
+  const records = await model.find({ code: Array.from(codes) });
+  return keyBy(records, 'code');
+}
+
+async function getEntitiesByCode(models, responses) {
+  const entityCodes = new Set();
+  responses.forEach(({ entity_code: entityCode }) => entityCode && entityCodes.add(entityCode));
+  return getRecordsByCode(models.entity, entityCodes);
+}
+
+async function getQuestionsByCode(models, responses) {
+  const questionCodes = new Set();
+  responses.forEach(r => Object.keys(r.answers).forEach(code => questionCodes.add(code)));
+  return getRecordsByCode(models.question, questionCodes);
+}
+
+const UpsertAnswersSchema = yup.array().of(
+  yup.object({
+    id: yup.string().required,
+    type: yup.string().required,
+    question_id: yup.string().required,
+  }),
+);
+
+async function saveAnswerRecords(models, rawAnswerRecords, surveyResponseId, questionsByCode) {
+  let answers;
+  /** The format of answers between upserting and creating are different.
+   *  create answer records: [ { [questionCode]: [answerValue] } ]
+   *  upsert answer records: [ UpsertAnswersSchema ]
+   */
```

Comment:
Combining upsert and create survey response logic is great, but it is a complicated when it comes to different formats of answer records...